packages/vhd-lib/parseVhdStream.js

```diff
@@ -13,124 +14,249 @@ const cappedBufferConcat = (buffers, maxSize) => {
+              type: 'block',
+              id: blockCounter,
+              offset: batEntryBytes,
+              size: fullBlockSize,
+            })
+            blockCount++
+          }
         }
+        // sort again index to ensure block and parent locator are in the right order
+        this._index.sort((a, b) => a.offset - b.offset)
+        item.blockCount = blockCount
+        batFound = true
       }
-      // sort again index to ensure block and parent locator are in the right order
-      index.sort((a, b) => a.offset - b.offset)
-      item.blockCount = blockCount
-    } else if (type === 'block') {
-      item.bitmap = buffer.slice(0, bitmapSize)
-      item.data = buffer.slice(bitmapSize)
+      yield item
     }
+  }
+
+  async *blocks() {
+    while (this._index.length > 0) {
+      const item = this._index.shift()
+      const buffer = await this._read(item.offset, item.size)
 
-    yield item
+      item.bitmap = buffer.slice(0, this._bitmapSize)
+      item.data = buffer.slice(this._bitmapSize)
+      item.buffer = buffer
+      yield item
+    }
+    /**
+     * the second footer is at filesize - 512 , there can be empty spaces between last block
+     * and the start of the footer
+     *
+     * we read till the end of the stream, and use the last 512 bytes as the footer
+     */
+    const bufFooterEnd = await readLastSector(this._stream)
+    assert(this.#bufFooter.equals(bufFooterEnd), 'footer1 !== footer2')
   }
 
-  /**
-   * the second footer is at filesize - 512 , there can be empty spaces between last block
-   * and the start of the footer
-   *
-   * we read till the end of the stream, and use the last 512 bytes as the footer
-   */
-  const bufFooterEnd = await readLastSector(stream)
-  assert(bufFooter.equals(bufFooterEnd), 'footer1 !== footer2')
+  async *parse() {
+    yield* this.headers()
+    yield* this.blocks()
+  }
 }
 
-function readLastSector(stream) {
-  return new Promise((resolve, reject) => {
-    let bufFooterEnd = Buffer.alloc(0)
-    stream.on('data', chunk => {
-      if (chunk.length > 0) {
-        bufFooterEnd = cappedBufferConcat([bufFooterEnd, chunk], SECTOR_SIZE)
+// hybrid mode : read the headers from the vhd stream, and read the blocks from nbd
+class StreamNbdParser extends StreamParser {
+  #nbdClient = null
+  #concurrency = 16
+
+  constructor(stream, nbdClient = {}) {
+    super(stream)
+    this.#nbdClient = nbdClient
+  }
+
+  async readBlockData(item) {
+    const SECTOR_BITMAP = Buffer.alloc(512, 255)
+    const client = this.#nbdClient
+    // we read in a raw file, so the block position is id x length, and have nothing to do with the offset
+    // in the vhd stream
+    const rawDataLength = item.size - SECTOR_BITMAP.length
+    let data = await client.readBlock(item.id, rawDataLength)
+
+    // end of file , non aligned vhd block
+    if (data.length < rawDataLength) {
+      data = Buffer.concat([data, Buffer.alloc(rawDataLength - data.length)])
```

Comment:
nbd client does not clip last block
Should check if it still works with unaligned vhd length